File: Python/25-reverse-nodes-in-k-group/index1.py
# Definition for singly-linked list.
# class ListNode:
#   def __init__(self, x):
#     self.val = x
#     self.next = None

class Solution:
  def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
    if not head: return None

    count, node = 0, head
    while node and count < k:
      node = node.next
      count += 1

    if count < k: return head

    prev = None
    cur = head
    while count > 0:
      cur.next, prev, cur = prev, cur, cur.next
      count -= 1

    new_start = cur
    head.next = self.reverseKGroup(new_start, k)

    return prev

File: JavaScript/25-reverse-nodes-in-k-group/index1.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *   this.val = val;
 *   this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
const reverseKGroup = function(head, k) {
  let count = 0;
  let node = head;
  while (node && count < k) {
    node = node.next;
    count += 1;
  }

  if (count < k) return head;

  let prev = null;
  let cur = head;
  while (count > 0) {
    let temp = cur.next;
    cur.next = prev;
    prev = cur;
    cur.next = temp;

    count -= 1;
  }

  head.next = reverseKGroup(cur, k);

  return prev;
}